src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineTraceForestBuilder.kt

```diff
@@ -0,0 +1,258 @@
+import com.intellij.util.ui.JBUI
+import com.intellij.xdebugger.frame.XExecutionStack
+import com.intellij.xdebugger.frame.XStackFrame
+import com.nikitanazarov.coroutinestacks.ui.ContainerWithEdges
+import com.nikitanazarov.coroutinestacks.ui.ForestLayout
+import com.nikitanazarov.coroutinestacks.ui.Separator
+import com.sun.jdi.Location
+import org.jetbrains.kotlin.analysis.decompiler.stub.file.ClsClassFinder
+import org.jetbrains.kotlin.idea.debugger.coroutine.data.CoroutineInfoData
+import org.jetbrains.kotlin.idea.debugger.coroutine.data.CoroutineStackFrameItem
+import org.jetbrains.kotlin.idea.debugger.coroutine.data.State
+import org.jetbrains.kotlin.idea.debugger.coroutine.util.CoroutineFrameBuilder
+import java.awt.*
+import java.awt.event.MouseAdapter
+import java.awt.event.MouseEvent
+import java.util.*
+import javax.swing.border.Border
+import javax.swing.border.LineBorder
+
+data class Node(
+    val stackFrameItem: CoroutineStackFrameItem? = null,
+    var num: Int = 0, // Represents how many coroutines have this frame in their stack trace
+    val children: MutableMap<Location, Node> = mutableMapOf(),
+    var coroutinesActive: String = ""
+)
+
+data class CoroutineTrace(val stackFrameItems: MutableList<CoroutineStackFrameItem?>, val header: String, val coroutinesActiveLabel: String)
+
+private fun CoroutineInfoData.render(): String =
+    "${descriptor.name}${descriptor.id} ${descriptor.state}\n"
+
+fun SuspendContextImpl.buildCoroutineStackForest(
+    rootValue: Node,
+    coroutineDataList: MutableList<CoroutineInfoData>
+): JBScrollPane? {
+    val lastRunningStackFrame = buildStackFrameGraph(coroutineDataList, rootValue)
+    val coroutineTraces = createCoroutineTraces(rootValue)
+    return createCoroutineTraceForest(coroutineTraces, lastRunningStackFrame)
+}
+
+private fun SuspendContextImpl.createCoroutineTraceForest(
+    traces: List<CoroutineTrace?>,
+    lastRunningStackFrame: String
+): JBScrollPane? {
+    if (traces.isEmpty()) return null
+    val vertexData: MutableList<JBList<String>?> = mutableListOf()
+    val componentData: MutableList<Component> = mutableListOf()
+    var previousListSelection: JBList<*>? = null
+    var maxWidth = 0
+    var traceNotNullCount = 0
+
+    traces.forEach { trace ->
+        if (trace == null) {
+            vertexData.add(null)
+            return@forEach
+        }
+
+        val vertex = CoroutineTraceUI(trace, lastRunningStackFrame, this).getVertex()
+        vertex.addListSelectionListener { e ->
+            val currentList = e.source as? JBList<*> ?: return@addListSelectionListener
+            if (previousListSelection != currentList) {
+                previousListSelection?.clearSelection()
+            }
+            previousListSelection = currentList
+        }
+        vertexData.add(vertex)
+        maxWidth += vertex.preferredWidth
+        traceNotNullCount += 1
+    }
+
+    val averagePreferredWidth = maxWidth / traceNotNullCount
```

Comment:
what if `traceNotNullCount` is zero?

Reply:
This can only happen if the `traces` list is ill-formed and only contains nulls, but we have to check for this situation anyway for our code to be safe